### PR TITLE
feat: add UltraChat 200k dataset support

### PR DESCRIPTION
**Pull request by omarfathy7**
## Summary

adds built-in UltraChat 200k (`HuggingFaceH4/ultrachat_200k`) dataset support for multi-turn conversational language modeling

addresses part of #125

### key implementation details

- **transform:** formats messages as `User: <content>\n` / `Assistant: <content>\n` for causal language modeling
- **evaluation:** preserves multi-turn context up to the final user turn and ends with `Assistant: ` user-ended conversations keep the final message
- **splits:** maps `train_sft` to training and `test_sft` to validation `train_gen` / `test_gen` are out of scope
- **config & registry:** adds typed configs and registers the dataset with the GPT-2 tokenizer preprocessor
- **documentation:** adds the UltraChat 200k dataset template

## AI usage

- [ ] No AI tools were used for this contribution
- [x] AI tools were used to assist this contribution

**Tool(s) and usage:**

ai assistants (`ChatGPT`, `Antigravity`, `ZCode`) were used during development for implementation support, debugging, and code review

## Testing

- [x] Tests added or updated where needed
- [x] Existing tests and checks pass
- [x] Documentation updated if needed

### verification details

- **unit tests:** 5/5 UltraChat tests passed
- **full test suite:** 138 tests passed 1 deselected
- **Ruff:** `uvx ruff check trainite/ tests/` passed
- **pre-commit:** `uv run pre-commit run --all-files` passed including Ruff formatting and Pyrefly type checking
- **cli generation:** `trainite init my-chat-model --dataset ultrachat-200k` successfully generated the project and config with `train_sft` and `test_sft`
- **real dataset:** successfully loaded UltraChat 200k from Hugging Face
  - `train_sft`: 207865 examples
  - `test_sft`: 23110 examples
- **training integration:** the generated project entered the real training loop and ran for 330 iterations the logged loss decreased from `10.9655` to `9.2114` before the run was interrupted
- **manual validation:** verified the transform with the same GPT-2 tokenizer used by Trainite including multi-turn formatting causal token alignment evaluation prompt construction and user-ended conversations

### CODE REVIEW

**Comment by aaishwarymishra** [comment]
@TahaZahid05 can you review it when you get time :)

File: trainite/datasets/ultrachat_200k.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+from typing import Any
+
+import torch
+from pydantic import BaseModel, ConfigDict
+
+
+class DatapointModel(BaseModel):
+    """Tokenized UltraChat 200k sample used by Trainite's causal LM pipeline."""
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]: This docstring is not descriptive enough as it doesn't explain the contract to a user reading it. Refer to other dataset implementations to get a more descriptive docstring: https://github.com/pytorch-ignite/trainite/blob/84f4dd9641f0785991b51eea3e4d7bc7261aeea3/trainite/datasets/string_reverse.py#L129-L135

File: trainite/datasets/ultrachat_200k.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    source: str
+    target: str
+    train_input_ids: torch.Tensor
+    train_label_ids: torch.Tensor
+    attention_mask: torch.Tensor
+    eval_input_ids: torch.Tensor
+
+
+class UltraChat200kTransform:
+    """Convert an UltraChat 200k sample into a causal language-modeling datapoint.
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]: This is too minimal. The transform required for ultrachat requires a much clearer docstring to explain to the user what it does.

File: trainite/datasets/ultrachat_200k.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        )
+
+        if messages and str(messages[-1].get("role", "")).strip().lower() == "assistant":
+            eval_messages = messages[:-1]
+            target = str(messages[-1].get("content", "")).strip()
+        else:
+            eval_messages = messages
+            target = ""
+
+        eval_prompt = f"{self._render_messages(eval_messages)}Assistant: "
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]: should we put the space after the colon or not? Will the tokenization differ for the model predicting ` Paris` compared to `Paris`? How will the GPT -2 tokenizer tokenize the space?

File: trainite/datasets/ultrachat_200k.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    def __call__(self, sample: dict[str, Any]) -> DatapointModel:
+        messages = sample.get("messages")
+
+        if not isinstance(messages, list):
+            raise ValueError("messages must be a list")
+
+        for message in messages:
+            if not isinstance(message, dict):
+                raise ValueError("each message in messages must be a dict")
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py:88]: what if `messages = []` ? We need to handle that case as well.

File: trainite/datasets/ultrachat_200k.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        attention_mask = tokenized["attention_mask"]
+
+        train_input_ids = torch.tensor(
+            token_ids[:-1],
+            dtype=torch.long,
+        )
+        train_label_ids = torch.tensor(
+            token_ids[1:],
+            dtype=torch.long,
+        )
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py:113]: shouldn't we mask user prompt tokens? Why do we need train the model to predict user's tokens?

**Comment by TahaZahid05** [comment]
@omarfathy7 Hi! You can contact me on Discord if you are facing any confusion regarding the PR.

**Comment by omarfathy7** [comment]
> @omarfathy7 Hi! You can contact me on Discord if you are facing any confusion regarding the PR.

hey @TahaZahid05  sorry for the late reply
ive been really busy with two other hackathons and submissions just started so i havent finished everything yet
sorry i didnt let you know i'd be late i just got caught up with everything
i'll send the updates by the end of the day and if theres any issue i'll let you know on discord
thanks for checking in
hope you understand

**Comment by TahaZahid05** [comment]
@omarfathy7 Hi! It's totally fine. I was just checking in to make sure that the PR itself wasn't causing any issues for you. Take your time and get back to it when you can, no rush!

Thanks for replying!

File: trainite/datasets/ultrachat_200k.py (hunk 1 above)
Comment (omarfathy7) [reply trainite/datasets/ultrachat_200k.py]: updated the docstring to follow the standard contract convention from `DatapointModel` in `string_reverse.py` explaining the role of `train_*`, `attention_mask`, `eval_input_ids`, and `source`/`target` fields

File: trainite/datasets/ultrachat_200k.py (hunk 2 above)
Comment (omarfathy7) [reply trainite/datasets/ultrachat_200k.py]: expanded the docstring following the pattern in `PromptCompletionTransform` in `string_reverse.py` it now explains the multi-turn rendering format teacher-forcing shifting supervised loss masking on assistant content and the evaluation prompt structure

File: trainite/datasets/ultrachat_200k.py (hunk 3 above)
Comment (omarfathy7) [reply trainite/datasets/ultrachat_200k.py]: in BPE tokenizers such as GPT-2 whitespace can affect tokenization and `" Paris"` can map to a different token ID than `"Paris"` without the space a trailing space after Assistant: can also produce a separate space token

the evaluation prompt now ends strictly with `Assistant:` without trailing whitespace

i also learned from this comment that token IDs exist and are different from tokens

File: trainite/datasets/ultrachat_200k.py (hunk 4 above)
Comment (omarfathy7) [reply trainite/datasets/ultrachat_200k.py:88]: added validation to reject empty lists with `raise ValueError("messages must not be empty")`, alongside validation for non-dict items added unit test coverage for this edge case in `tests/datasets/ultrachat_200k_test.py`

File: trainite/datasets/ultrachat_200k.py (hunk 5 above)
Comment (omarfathy7) [reply trainite/datasets/ultrachat_200k.py:113]: implemented standard conversational SFT loss masking `train_label_ids` is initialized with `ignore_index` (`-100`) across all user tokens and role headers (`User:` and `Assistant:`), unmasking only the assistant reply tokens so the model is strictly supervised on generating assistant responses

unit tests have been updated with exact label masking assertions

**Comment by TahaZahid05** [comment]
@omarfathy7 Thanks for the update to the PR! Can you resolve conflicts?

**Comment by omarfathy7** [comment]
> @omarfathy7 Thanks for the update to the PR! Can you resolve conflicts?

@TahaZahid05  done, thank you for the review and helpful feedback

File: trainite/datasets/ultrachat_200k.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+class DatapointModel(BaseModel):
+    """Contract for a causal-LM transformed item: training tensors + the eval prompt.
+
+    Convention: every causal-LM dataset transform returns this shape. The collate
+    fn batches the `train_*`/`attention_mask` fields; the trainer's inference loop
+    reads `eval_input_ids`/`source`/`target` directly.
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]:
```suggestion
```

File: trainite/datasets/ultrachat_200k.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
+from typing import Any
+
+import torch
+from pydantic import BaseModel, ConfigDict
+
+
+class DatapointModel(BaseModel):
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]: let's move the class after `UltraChat200kTransform`

File: trainite/datasets/ultrachat_200k.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    UltraChat 200k samples contain a ``messages`` list of role/content dictionaries
+    representing multi-turn dialogues. Each turn is rendered into standard plaintext
+    as ``User: <content>\\n`` or ``Assistant: <content>\\n``.
+
+    Token layout & supervised loss masking:
+        In conversational SFT (Supervised Fine-Tuning), the model is trained to predict
+        assistant responses given the conversational history, without computing loss on
+        user queries or role headers.
+
+        input_ids : full conversation tokens shifted right by 1 (= token_ids[:-1])
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]:
```suggestion
        input_ids : all conversation tokens except the last one (= token_ids[:-1])
```

File: trainite/datasets/ultrachat_200k.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+        labels    : full conversation tokens shifted left by 1 (= token_ids[1:]),
+                    where all user message tokens and role headers are masked with ``ignore_index`` (-100).
+                    Only assistant response tokens are active loss targets.
+
+    During evaluation (inference):
+        The evaluation prompt preserves conversational context up to the final user turn and
+        ends with ``Assistant:`` (without trailing whitespace to prevent BPE tokenization artifacts),
```
Comment (TahaZahid05) [inline trainite/datasets/ultrachat_200k.py]:
```suggestion
        ends with ``Assistant:``,
```

File: trainite/templates/components/datasets/ultrachat_200k.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      max_length: 128
+      ignore_index: -100
+```
+
+`max_length` controls the maximum sequence length, and `ignore_index` sets the label index for masked tokens (defaults to `-100`).
+
+The transform renders each turn as `User: <content>\n` and `Assistant: <content>\n`.
+The tokenizer handles special-token addition. The transform creates:
+
+* `train_input_ids`: Full sequence tokens shifted right by 1 (`token_ids[:-1]`).
```
Comment (TahaZahid05) [inline trainite/templates/components/datasets/ultrachat_200k.md]:
```suggestion
* `train_input_ids`: All sequence tokens except the last one (`token_ids[:-1]`).
```

File: trainite/templates/components/datasets/ultrachat_200k.md [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+* `train_label_ids`: Full sequence tokens shifted left by 1 (`token_ids[1:]`), with user prompt tokens and role headers masked to `ignore_index` (`-100`) so loss is only calculated on assistant responses.
+* `attention_mask`: Attention mask for the input sequence (`attention_mask[:-1]`).
+* `eval_input_ids`: Generation prompt token IDs ending with `Assistant:` (without trailing whitespace).
```
Comment (TahaZahid05) [inline trainite/templates/components/datasets/ultrachat_200k.md]:
```suggestion
* `eval_input_ids`: Generation prompt token IDs ending with `Assistant:`.
```

File: trainite/templates/components/datasets/ultrachat_200k.md [hunk 12]
```diff
@@ -0,0 +1,57 @@
+# UltraChat 200k dataset
+
+This built-in dataset loads the UltraChat 200k dataset from Hugging Face
+Datasets and applies a multi-turn conversation transform for causal language-modeling.
```
Comment (TahaZahid05) [inline trainite/templates/components/datasets/ultrachat_200k.md]:
```suggestion
This built-in dataset loads the ['HuggingFaceH4/ultrachat_200k'](https://huggingface.co/datasets/HuggingFaceH4/ultrachat_200k) dataset from Hugging Face and applies a multi-turn conversation transform for causal language-modeling.
```

File: trainite/templates/components/datasets/ultrachat_200k.md [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+The transform renders each turn as `User: <content>\n` and `Assistant: <content>\n`.
+The tokenizer handles special-token addition. The transform creates:
+
+* `train_input_ids`: Full sequence tokens shifted right by 1 (`token_ids[:-1]`).
+* `train_label_ids`: Full sequence tokens shifted left by 1 (`token_ids[1:]`), with user prompt tokens and role headers masked to `ignore_index` (`-100`) so loss is only calculated on assistant responses.
+* `attention_mask`: Attention mask for the input sequence (`attention_mask[:-1]`).
+* `eval_input_ids`: Generation prompt token IDs ending with `Assistant:` (without trailing whitespace).
+
+During evaluation and inference logging, the prompt preserves the multi-turn context and ends with
+`Assistant:` so the model generates the final assistant response.
```
Comment (TahaZahid05) [inline trainite/templates/components/datasets/ultrachat_200k.md]: There is no need to go to a new line to complete one sentence. Kindly fix this by writing it in a single line instead.

File: trainite/templates/components/datasets/ultrachat_200k.md [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+
+## Dataset splits
+
+The `train_sft` split is used for training and `test_sft` is mapped to Trainite's validation split.
+The generation-ranking splits (`train_gen` and `test_gen`) are outside the scope of this built-in wrapper.
+
+## Hugging Face authentication
+
+The dataset is downloaded and cached by Hugging Face Datasets. Do not put Hugging Face access
+tokens in `config.yaml`. Authenticate through the Hugging Face CLI or the environment when required.
```
Comment (TahaZahid05) [inline trainite/templates/components/datasets/ultrachat_200k.md]: Same here. No need to go to a new line?

**Review by TahaZahid05 — COMMENTED** [review]
@omarfathy7  Some minor changes. We can merge after this!

**Comment by omarfathy7** [comment]
@TahaZahid05  thank you, i’ve tired you out😅

**Review by TahaZahid05 — APPROVED** [review]
@omarfathy7 No worries! I have reverted the Datapointmodel change since the typehint of the function looks more clean, and also added backtick rather than single quotes in documentation. Thanks!

**Comment by omarfathy7** [comment]
@TahaZahid05 do u recommend any other issues?
